src/binder/statement/create_table.rs

```diff
@@ -45,57 +45,73 @@ impl Binder {
+                    let mut is_primary = false;
+                    let mut col = ColumnCatalog::from(colum_def, &mut is_primary);
+                    if flag && is_primary {
+                        // we can't support sql query like
+                        // "create table (a int primary key, b int not null, primary key(b));"
+                        // declaring pk both in column's option and 'primary key(c1..)' syntax"
+                        return Err(BindError::NotSupportedTSQL);
+                    } else if is_primary {
+                        // primary key is declared only in column's options
+                        ordered_pk_ids.push(idx as ColumnId);
+                    } else if ordered_pk_pos.contains_key(col.name()) {
+                        // primary key is declared only by "primary key(c1, c2) syntax"
+                        let pos = *(ordered_pk_pos.get(col.name()).unwrap());
+                        ordered_pk_ids[pos] = idx as ColumnId;
+                        col.set_primary(true); // TODO: remove this line in the future
                     }
+                    col.set_id(idx as ColumnId);
+                    col_catalogs.push(col);
                 }
 
-                let columns = columns
-                    .iter()
-                    .enumerate()
-                    .map(|(idx, col)| {
-                        let mut col = ColumnCatalog::from(col);
-                        if extra_pk_name.contains(col.name()) && !col.is_primary() {
-                            col.set_primary(true);
-                        }
-                        col.set_id(idx as ColumnId);
-                        col
-                    })
-                    .collect();
                 Ok(BoundCreateTable {
                     database_id: db.id(),
                     schema_id: schema.id(),
                     table_name: table_name.into(),
-                    columns,
+                    columns: col_catalogs,
+                    ordered_pk_ids,
                 })
             }
             _ => panic!("mismatched statement type"),
         }
     }
+
+    fn get_extra_pks(constraints: &Vec<TableConstraint>) -> HashMap<String, usize> {
+        let mut ordered_pk_pos: HashMap<String, usize> = HashMap::new();
+        for constraint in constraints {
+            match constraint {
+                TableConstraint::Unique {
+                    is_primary,
+                    columns,
+                    ..
+                } if *is_primary => columns.iter().enumerate().for_each(|(index, indent)| {
+                    ordered_pk_pos.insert(indent.value.clone(), index);
+                }),
+                _ => todo!(),
+            }
+        }
+        ordered_pk_pos
+    }
 }
 
-impl From<&ColumnDef> for ColumnCatalog {
-    fn from(cdef: &ColumnDef) -> Self {
+impl ColumnCatalog {
+    fn from(cdef: &ColumnDef, is_primary_: &mut bool) -> Self {
```

Comment:
Generally in the Rust programming language, `from` will only take one parameter. I'd suggest not doing two things at once in `from`. Just declare it as

`from(cdef: &ColumnDef) -> Self`

and access `is_primary` field for this column later.